pygem/cad/ffd.py

```diff
@@ -0,0 +1,353 @@
+        compound = TopoDS_Compound()
+        compound_builder.MakeCompound(compound)
+
+
+        # cycle on the faces to get the control points
+        # init some quantities
+        faceCount = 0
+        face_list = []
+        faces_explorer = TopExp_Explorer(shape, TopAbs_FACE)
+
+        while faces_explorer.More():
+            # performing some conversions to get the right
+            # format (BSplineSurface)
+            face = topods_Face(faces_explorer.Current())
+            nurbs_converter = BRepBuilderAPI_NurbsConvert(face)
+            nurbs_face = nurbs_converter.Shape()
+            face_aux = topods_Face(nurbs_face)
+            brep_face = BRep_Tool.Surface(topods_Face(nurbs_face))
+            bounds = 0.0
+            bounds = brep_face.Bounds()
+
+            bspline_face = geomconvert_SurfaceToBSplineSurface(brep_face)
+            # we will then add an amount of nodes that will grant
+            # us our prescribed resolution both along u and v
+            for i in range(self.uKnotsToAdd):
+                bspline_face.InsertUKnot(bounds[0]+ \
+                    i*(bounds[1]-bounds[0])/self.uKnotsToAdd, 1, self.tolerance)
+            for i in range(self.vKnotsToAdd):
+                bspline_face.InsertVKnot(bounds[2]+ \
+                    i*(bounds[3]-bounds[2])/self.vKnotsToAdd, 1, self.tolerance)
+
+            # openCascade object
+            occ_face = bspline_face
+
+            # extract the Control Points of each face
+            n_poles_u = occ_face.NbUPoles()
+            n_poles_v = occ_face.NbVPoles()
+            control_polygon_coordinates = np.zeros(\
+            shape=(n_poles_u * n_poles_v, 3))
+            # cycle over the poles to get their coordinates
+            i = 0
+            for pole_u_direction in range(n_poles_u):
+                for pole_v_direction in range(n_poles_v):
+                    control_point_coordinates = occ_face.Pole(\
+                    pole_u_direction + 1, pole_v_direction + 1)
+                    control_polygon_coordinates[i, :] = \
+                          [control_point_coordinates.X(),\
+                           control_point_coordinates.Y(),\
+                           control_point_coordinates.Z()]
+                    i += 1
+
+            ## SURFACES PHASE ###############################################
+            src_pts = control_polygon_coordinates
+            new_pts = super().__call__(src_pts) # dont touch this line
+
+            i = 0
+            for pole_u_direction in range(n_poles_u):
+                for pole_v_direction in range(n_poles_v):
+                    control_point = gp_Pnt(new_pts[i, 0],
+                                           new_pts[i, 1],
+                                           new_pts[i, 2])
+                    occ_face.SetPole(pole_u_direction + 1,
+                                     pole_v_direction + 1,
+                                     control_point)
+                    i += 1
```

Comment:
Class private method (one underscore in front of the name e.g. `_foo()`). Describe input and output variables.